src/Microsoft.Azure.WebJobs.Host/Bindings/BindingBase.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Reflection;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Host.Protocols;
+
+namespace Microsoft.Azure.WebJobs.Host.Bindings
+{
+
+    // Helper class for implementing IBinding with the attribute resolver pattern. 
+    internal abstract class BindingBase<TAttribute> : IBinding
+        where TAttribute : Attribute
+    {
+        protected readonly AttributeCloner<TAttribute> _cloner;
+        private readonly ParameterDescriptor _param;
+
+        public BindingBase(AttributeCloner<TAttribute> cloner, ParameterDescriptor param)
+        {
+            _cloner = cloner;
+            _param = param;
+        }
+
+        public BindingBase(AttributeCloner<TAttribute> cloner, ParameterInfo parameterInfo)
+        {
+            _cloner = cloner;
+            _param = new ParameterDescriptor
+            {
+                Name = parameterInfo.Name,
+                DisplayHints = new ParameterDisplayHints
+                {
+                    Description = "value"
+                }
+            };
+        }
+
+        public bool FromAttribute
+        {
+            get
+            {
+                return true;
+            }
+        }
+
+        protected abstract Task<IValueProvider> BuildAsync(TAttribute attrResolved);
+
+        public Task<IValueProvider> BindAsync(BindingContext context)
+        {
+            var attrResolved = _cloner.ResolveFromBindings(context);
+            return BuildAsync(attrResolved);
+        }
+
+        public Task<IValueProvider> BindAsync(object value, ValueBindingContext context)
+        {
+            var str = value as string;
+            if (str != null)
+            {
+                // Called when we invoke from dashboard. 
+                // str --> attribute --> obj 
+                var resolvedAttr = _cloner.ResolveFromInvokeString(str);
```

Comment:
If you need full control, the attribute can impl from IAttributeInvokeDescriptor and implement the conversions itself. 

If the attr doesn't impl that interface, then we infer a default implementation of the interface that:
1. picks the [Resolve] property 
2. is round-tripable.  (so replay works) 
The default is good enough for [Blob].


---
In reply to: [61131733](https://github.com/Azure/azure-webjobs-sdk/pull/694#discussion_r61131733) [](ancestors = 61131733)